types/ember-data/index.d.ts

```diff
@@ -119,11 +119,14 @@ export namespace DS {
     ): string;
     const VERSION: string;
 
-    interface AttrOptions<T extends Transform | string | number | boolean | Date = any> {
-        defaultValue?: T extends Transform ? (() => ReturnType<T['deserialize']>) : string | number | boolean | (() => T);
+    interface AttrOptions<T> {
+        defaultValue?: T extends Exclude<object, null> ? (() => T) : T | (() => T);
         allowNull?: boolean; // TODO: restrict to boolean transform (TS 2.8)
     }
 
+    // The TransformRegistry should really only contain transforms, but historically people have just put the return type directly in.
+    type TransformType<K extends keyof TransformRegistry> = TransformRegistry[K] extends Transform ? ReturnType<TransformRegistry[K]['deserialize']> : TransformRegistry[K];
```

Comment:
Gave people the option either way, nice!

Reply:
Can you add a test that uses the registry the "old way"?

Reply:
Ah, good call. I tested it in my app but forgot to add the test here.